### PR TITLE
96 ticker calculation

### PR DESCRIPTION
**Pull request by vstavetski**
Add ticker calculation functionality and stream api for it

### CODE REVIEW

File: stream-api/src/main/kotlin/fund/cyber/markets/api/common/BroadcastersIndex.kt [hunk 1]
```diff
@@ -1,63 +1,75 @@
...
-                    .map { (_, broadcaster) -> broadcaster }
-            else -> index
-                    .flatMap { (_, pairIndex) -> pairIndex.entries }
-                    .map { (_, broadcaster) -> broadcaster }
+    protected val index: MutableMap<BroadcasterDefinition, B> = ConcurrentHashMap()
+
+    fun broadcastersFor(pairInitializers: List<TokensPairInitializer>, exchanges: List<String>, windowDurations: List<Long> = emptyList()): Collection<B> {
+
+        var broadcasters : MutableMap<BroadcasterDefinition, B> = ConcurrentHashMap(index)
+        when {
```
Comment (hleb-albau) [inline stream-api/src/main/kotlin/fund/cyber/markets/api/common/BroadcastersIndex.kt]: wrong when

File: build.gradle [hunk 2]
```diff
@@ -159,3 +159,15 @@ project(":stream-processing") {
         compileOnly("org.apache.spark:spark-sql-kafka-0-10_2.11:$sparkVersion")
```
Comment (hleb-albau) [inline build.gradle]: please, remove this spark dependencies

File: settings.gradle [hunk 3]
```diff
@@ -2,3 +2,4 @@ include("core")
 include("stream-api")
 include("connectors")
 include("stream-processing")
```
Comment (hleb-albau) [inline settings.gradle]: remove stream processing module

File: tickers/src/main/kotlin/fund/cyber/markets/tickers/TickersApplication.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.kafka.common.serialization.Serdes
+import org.apache.kafka.streams.KafkaStreams
+import org.apache.kafka.streams.kstream.KStream
+import org.apache.kafka.streams.kstream.KStreamBuilder
+import org.apache.kafka.streams.kstream.TimeWindows
+import java.sql.Timestamp
+
+val configuration: KafkaConfiguration = KafkaConfiguration()
+
+fun main(args: Array<String>) {
```
Comment (hleb-albau) [inline tickers/src/main/kotlin/fund/cyber/markets/tickers/TickersApplication.kt]: just answer question: If error occurred inside stream handlers, will application exit?

File: tickers/src/main/kotlin/fund/cyber/markets/tickers/configuration/TickersKafkaConfiguration.kt [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+class KafkaConfiguration(
+        val kafkaServers: String = env(Constants.KAFKA_CONNECTION, "localhost:9092"),
+        val topicNamePattern: Pattern = Constants.TRADES_TOPIC_NAME_PATTERN,
+        val topicResubscribe: Long = TimeUnit.MINUTES.toMillis(1),
+        val windowDurationsString: String = env(Constants.WINDOW_DURATIONS_MIN, "1,5,15,30,60,180,240,360,720,1440"),
+        val windowHop: Long = env(Constants.WINDOW_HOP_SEC, 3000)
+) {
+
+    fun windowDurations() : List<Long> {
+       return windowDurationsString.split(",").map { it -> it.toLong()*60*1000 }
```
Comment (hleb-albau) [inline tickers/src/main/kotlin/fund/cyber/markets/tickers/configuration/TickersKafkaConfiguration.kt]: formating -> control + alt + shift + L

File: .circleci/config.yml [hunk 6]
```diff
@@ -62,6 +62,26 @@ jobs:
             - ~/.gradle/caches/
```
Comment (hleb-albau) [inline .circleci/config.yml]: right now deploy_kafka-connectors_image is broken, please, fix it

File: .circleci/config.yml [hunk 7]
```diff
@@ -62,6 +62,26 @@ jobs:
...
             - ~/.gradle/wrapper/
 
+  build_tickers:
+    <<: *default_container
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
```
Comment (hleb-albau) [inline .circleci/config.yml]: should not be api-dependencies

File: tickers/src/main/kotlin/fund/cyber/markets/tickers/TickersApplication.kt [hunk 8]
```diff
@@ -9,20 +9,26 @@ import fund.cyber.markets.tickers.configuration.tickersTopicName
...
 
 
 fun main(args: Array<String>) {
 
     val configuration = KafkaConfiguration()
 
-    val builder = KStreamBuilder()
-    val tradeStream = builder.stream<String, Trade>(Serdes.String(), JsonSerde(Trade::class.java), configuration.topicNamePattern)
+    val builder = StreamsBuilder()
+    val tradeStream = builder.stream<String, Trade>(configuration.topicNamePattern, Consumed.with(Serdes.String(), JsonSerde(Trade::class.java)))
```
Comment (hleb-albau) [inline tickers/src/main/kotlin/fund/cyber/markets/tickers/TickersApplication.kt]: just answer question: new topics re-subscription mechanism enabled?

File: core/src/main/kotlin/fund/cyber/markets/kafka/Serialization.kt [hunk 9]
```diff
@@ -21,9 +21,12 @@ class JsonDeserializer<T>(private val type: Class<T>) : Deserializer<T> {
     private val objectMapper = ObjectMapper().registerKotlinModule()
     private val LOGGER = LoggerFactory.getLogger(JsonDeserializer::class.java)!!
 
-    override fun deserialize(topic: String, data: ByteArray): T {
+    override fun deserialize(topic: String, data: ByteArray?): T? {
         try {
-            return objectMapper.readValue(data, type)
+            return if (data == null)
```
Comment (hleb-albau) [inline core/src/main/kotlin/fund/cyber/markets/kafka/Serialization.kt]: can be simplified (kotlin lang)

File: tickers/src/main/kotlin/fund/cyber/markets/tickers/TickersApplication.kt (hunk 4 above)
Comment (vstavetski) [reply tickers/src/main/kotlin/fund/cyber/markets/tickers/TickersApplication.kt]: yes, the application will exit if any exception occurs

**Pull request merged**